crates/chia-consensus/src/gen/solution_generator.rs

```diff
@@ -39,36 +40,41 @@ where
     Ok(quote)
 }
 
+// do not use generally
+fn len_for_generator(val: u64) -> usize {
```

Comment:
I don't quite understand what this function is doing exactly. The return values aren't intuitive to me.
It would be easy to think that this function ought to return the same as:

```
let mut a = Allocator::new();
let atom = a.new_number(val);
a.atom_len(atom)
```

But you're primarily concerned about the *serialized* length, right? Which is slightly different, a value < `0x80` is serialized as a single byte, but as soon as you need two bytes, we also have a length prefix for the atom. So I think it's more accurate to say that this function should return the same as:

```
let mut a = Allocator::new();
let atom = a.new_number(val);
let bytes = node_to_bytes(&a, atom);
bytes.len()
```

Reply:
This is generally accurate, however I observed that the chialisp itself gets optimised when the size of the amount is less than 128, saving 1 byte. So while `len(bytes(Program.to(97)))` is actually 1, with the optimisation the net size change is 0.

Reply:
I think this is a misunderstanding. the length-prefix for atoms is similar to UTF-8, so small values (< 0x80) are encoded as a single byte (the leading 0 bit indicates a small value). I suspect your calculation compensates for this 0-bytes by adding it back later

Reply:
The full details of the serialization can be found here: https://chialisp.com/clvm/#large-atoms

In the case of a `u64`, we know that it will never be more than 9 bytes (including a leading zero if the value is large enough). That means the *length prefix* never has to be more than 1 byte.
If the value is < `0x80`, it's encoded as 1 byte. Otherwise it's encoded as 1-byte length-prefix (`0b10xxxxxx`) + the actual bytes themselves. The bytes being encoded as variable length, signed, big endian. I think you end up with something like this:

| value | serialized length |
| --- | --- |
| 0 | 1 byte |
| 0x7f | 1 byte |
| 0x80 | 1 byte prefix + 2 bytes value |
| 0x7fff | 1 byte prefix + 2 bytes value |
| 0x8000 | 1 byte prefix + 3 bytes value |
| 0x800000 | 1 byte prefix + 4 bytes value |

and so on. I think it would make it simpler to have this function be well defined as the serialized length of the `u64` value. You could then also have a test that covers the most relevant values (i.e. the boundary values where it the length changes)